### PR TITLE
fix(@angular/cli): only values in enum should be allowed to update

### PR DESCRIPTION
**Pull request by sumitarora**
Fixes: https://github.com/angular/angular-cli/issues/5874

### CODE REVIEW

File: packages/@ngtools/json-schema/src/schema-class-factory.ts [hunk 1]
```diff
@@ -132,7 +131,11 @@ class SchemaClassBase<T> implements SchemaClass<T> {
   /** Set a value from a JSON path. */
   $$set(path: string, value: any): void {
     const node = _getSchemaNodeForPath(this.$$schema(), path);
+
     if (node) {
+      if (node.constructor.name === 'EnumSchemaTreeNode' && node.items.indexOf(value) === -1) {
+        throw new InvalidUpdateValue('Invalid value can only be one of these: ' + node.items);
```
Comment (hansl) [inline packages/@ngtools/json-schema/src/schema-class-factory.ts]: Did you check what kind of output this generates? `items.toString()` doesn't seem to be what we want to show.

File: packages/@ngtools/json-schema/src/schema-class-factory.ts [hunk 2]
```diff
@@ -132,7 +131,11 @@ class SchemaClassBase<T> implements SchemaClass<T> {
   /** Set a value from a JSON path. */
   $$set(path: string, value: any): void {
     const node = _getSchemaNodeForPath(this.$$schema(), path);
+
     if (node) {
+      if (node.constructor.name === 'EnumSchemaTreeNode' && node.items.indexOf(value) === -1) {
```
Comment (clydin) [inline packages/@ngtools/json-schema/src/schema-class-factory.ts]: You could probably place this inside `EnumSchemaTreeNode` directly as an override of the `set` function.  No need for the constructor name checking in that case.

File: packages/@ngtools/json-schema/src/schema-class-factory.ts (hunk 1 above)
Comment (sumitarora) [inline packages/@ngtools/json-schema/src/schema-class-factory.ts]: @hansl Prints like this `Invalid value can only be one of these: Emulated,Native,None`

File: packages/@ngtools/json-schema/src/schema-class-factory.ts (hunk 2 above)
Comment (sumitarora) [inline packages/@ngtools/json-schema/src/schema-class-factory.ts]: @clydin done the changes 👍

**Comment by filipesilva** [comment]
@hansl can you re-review?

File: packages/@ngtools/json-schema/src/schema-tree.spec.ts [hunk 3]
```diff
@@ -55,12 +55,12 @@ describe('@ngtools/json-schema', () => {
...
       expect(proto.a instanceof Array).toBe(true);
-      expect(proto.a).toEqual([undefined, 'v1', undefined, 'v3']);
+      expect(proto.a).toEqual(['v1', 'v3']);
 
       // Set it to a string, which is valid.
       proto.a[0] = 'v2';
-      proto.a[1] = 'INVALID';
-      expect(proto.a).toEqual(['v2', undefined, undefined, 'v3']);
+      proto.a[1] = 'v3';
+      expect(proto.a).toEqual(['v2', 'v3']);
```
Comment (hansl) [inline packages/@ngtools/json-schema/src/schema-tree.spec.ts]: You're not testing anywhere that setting `INVALID` would throw an error. You should.

**Comment by hansl** [comment]
Need a unit test for the new use case.

**Comment by sumitarora** [comment]
@hansl Added test for new case.

**Pull request merged**